pilot/cmd/pilot-agent/status/server_test.go

```diff
@@ -255,57 +269,93 @@ func TestAppProbe(t *testing.T) {
-	server, err := NewServer(Config{
-		StatusPort: 0,
-		KubeAppProbers: fmt.Sprintf(`{"/app-health/hello-world/readyz": {"httpGet": {"path": "/hello/sunnyvale", "port": %v}},
-"/app-health/hello-world/livez": {"httpGet":{"port": %v}}}`, appPort, appPort),
-	})
-	if err != nil {
-		t.Errorf("failed to create status server %v", err)
-		return
-	}
-	go server.Run(context.Background())
-
-	var statusPort uint16
-	for statusPort == 0 {
-		server.mutex.RLock()
-		statusPort = server.statusPort
-		server.mutex.RUnlock()
+	simpleConfig := KubeAppProbers{
+		"/app-health/hello-world/readyz": &Prober{
+			HTTPGet: &v1.HTTPGetAction{
+				Path: "/hello/sunnyvale",
+				Port: intstr.IntOrString{IntVal: int32(appPort)},
+			},
+		},
+		"/app-health/hello-world/livez": &Prober{
+			HTTPGet: &v1.HTTPGetAction{
+				Port: intstr.IntOrString{IntVal: int32(appPort)},
+			},
+		},
 	}
-
-	t.Logf("status server starts at port %v, app starts at port %v", statusPort, appPort)
+	_ = simpleConfig
 	testCases := []struct {
 		probePath  string
+		config     KubeAppProbers
 		statusCode int
 	}{
 		{
-			probePath:  fmt.Sprintf(":%v/bad-path-should-be-404", statusPort),
+			probePath:  "bad-path-should-be-404",
+			config:     simpleConfig,
 			statusCode: http.StatusNotFound,
 		},
 		{
-			probePath:  fmt.Sprintf(":%v/app-health/hello-world/readyz", statusPort),
+			probePath:  "app-health/hello-world/readyz",
+			config:     simpleConfig,
 			statusCode: http.StatusOK,
 		},
 		{
-			probePath:  fmt.Sprintf(":%v/app-health/hello-world/livez", statusPort),
+			probePath:  "app-health/hello-world/livez",
+			config:     simpleConfig,
+			statusCode: http.StatusOK,
+		},
+		{
+			probePath: "app-health/header/readyz",
+			config: KubeAppProbers{
+				"/app-health/header/readyz": &Prober{
+					HTTPGet: &v1.HTTPGetAction{
+						Port: intstr.IntOrString{IntVal: int32(appPort)},
+						Path: "/header",
+						HTTPHeaders: []v1.HTTPHeader{
+							{"Host", "host"},
+							{"Some-Header", "some-value"},
+						},
+					},
+				},
+			},
 			statusCode: http.StatusOK,
 		},
 	}
 	for _, tc := range testCases {
-		client := http.Client{}
-		req, err := http.NewRequest("GET", fmt.Sprintf("http://localhost%s", tc.probePath), nil)
-		if err != nil {
-			t.Errorf("[%v] failed to create request", tc.probePath)
-		}
-		resp, err := client.Do(req)
-		if err != nil {
-			t.Fatal("request failed")
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode != tc.statusCode {
-			t.Errorf("[%v] unexpected status code, want = %v, got = %v", tc.probePath, tc.statusCode, resp.StatusCode)
-		}
+		t.Run(tc.probePath, func(t *testing.T) {
+			// Starts the pilot agent status server.
+			server, err := NewServer(Config{
+				StatusPort: 0,
+			})
+			if err != nil {
+				t.Fatalf("failed to create status server %v", err)
+			}
+			server.appKubeProbers = tc.config
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			_ = cancel
+			go server.Run(ctx)
```

Comment:
i think you are leaking the go routine in the test without explicitly shutdown the server? basically go.server.Run(ctx) would never return on its own? inside the `Serve` never returns unless error in the setup or `httpServer.Shutdown` is invoked

Reply:
That is what `defer cancel()` is for

Reply:
It will exit once the context is canceled  then we call `listener.close()`

Reply:
no. `listener.Close()` is in the outer Run function body, and http.Serve is kicked in another inner go function, Serve is a blocking function call. once go routine starts, nothing stops it other than process stops.

ctx.Done is invoked after the Serve(), which has no impact on http server being continue to run.

Reply:
original implementation is already like this, comment since this PR is around adjacent code area.

Reply:
I mean `l.Close()` inside server.go

Reply:
I am not sure a good way to actually verify goroutine leak in the go tests though.. there is probably a way to be sure

Reply:
sorry, i just forgot that when listener is closed the serve will also return. sorry about that.